cirq/circuits/circuit.py

```diff
@@ -1310,22 +1392,20 @@ def tetris_concat(
 
         # Accumulate all the circuits into the buffer.
         for k in range(1, len(circuits)):
-            offset, n_acc = _tetris_concat_helper(
-                offset, n_acc, buffer, circuits[k].moments, stop_at_first_alignment
-            )
+            offset, n_acc = _tetris_concat_helper(offset, n_acc, buffer, circuits[k].moments, align)
 
         return cirq.Circuit(buffer[offset : offset + n_acc])
 
 
 def _overlap_collision_time(
-    c1: Sequence['cirq.Moment'], c2: Sequence['cirq.Moment'], stop_at_first_alignment: bool
+    c1: Sequence['cirq.Moment'], c2: Sequence['cirq.Moment'], align: 'cirq.Alignment'
 ) -> int:
     # Tracks the first used moment index for each qubit in c2.
     # Tracks the complementary last used moment index for each qubit in c1.
     seen_times: Dict['cirq.Qid', int] = {}
 
     # Start scanning from end of first and start of second.
-    upper_bound = (min if stop_at_first_alignment else max)(len(c1), len(c2))
+    upper_bound = len(c1) if align == Alignment.START else len(c2)
```

Comment:
Note that in the case of Alignment.END, and len(c2) > len(c1), and they don't intersect, then this implementation pushes c2 to start before c1. 

If you want to avoid this behavior, we need `len(c1) if align == Alignment.START else min(len(c1), len(c2))`. This creates the behavior where:

* Alignment.START is tetris-style.
* Alignment.END is tetris-style *except* it doesn't allow the last moment of c2 to push past the last moment of c1. It ensures c2 stays at the end of the resulting circuit.

I *think* these are the two most likely required use cases.

Reply:
In that case I _would_ expect c2 to start before c1, since I specified that the ends should align. I think the code as-is handles all the cases:
```python
In [1]: import cirq

In [2]: a, b = [cirq.X(q) for q in cirq.LineQubit.range(2)]

In [3]: cirq.Circuit.tetris_concat(cirq.Circuit([a] * 5), cirq.Circuit([b] * 3), align='END')
Out[3]: 
0: ───X───X───X───X───X───

1: ───────────X───X───X───

In [4]: cirq.Circuit.tetris_concat(cirq.Circuit([a] * 5), cirq.Circuit([b] * 3), align='START')
Out[4]: 
0: ───X───X───X───X───X───

1: ───X───X───X───────────

In [5]: cirq.Circuit.tetris_concat(cirq.Circuit([a] * 3), cirq.Circuit([b] * 5), align='START')
Out[5]: 
0: ───X───X───X───────────

1: ───X───X───X───X───X───

In [6]: cirq.Circuit.tetris_concat(cirq.Circuit([a] * 3), cirq.Circuit([b] * 5), align='END')
Out[6]: 
0: ───────────X───X───X───

1: ───X───X───X───X───X───
```

Reply:
+1 to @maffoo's direction here - I think having a "tetris" stop always at the start is more confusing to think about. Aligning ends or starts is just very simple.